### PR TITLE
Add Hierarchical Continuous Aggregates validations

### PR DESCRIPTION
**Pull request by fabriziomello**
Commit https://github.com/timescale/timescaledb/commit/3749953e9704e45df8f621607989ada0714ce28d introduce Hierarchical Continuous Aggregates (aka
Continuous Aggregate on top of another Continuous Aggregate) but it
lacks of some basic validations.

Validations added during the creation of a Hierarchical Continuous
Aggregate:

* Forbid create a continuous aggregate with fixed-width bucket on top of
  a continuous aggregate with variable-width bucket.

* Forbid incompatible bucket widths:
  - should not be equal;
  - bucket width of the new continuous aggregate should be greater than
    the source continuous aggregate;
  - bucket width of the new continuous aggregate should be multiple of
    the source continuous aggregate.

### CODE REVIEW

File: src/ts_catalog/continuous_agg.c [hunk 1]
```diff
@@ -1898,3 +1899,38 @@ ts_cagg_permissions_check(Oid cagg_oid, Oid userid)
...
+		cagg_view_oid =
+			get_relname_relid(NameStr(cagg->data.user_view_name),
+							  get_namespace_oid(NameStr(cagg->data.user_view_schema), false));
+
+	cagg_view_rel = table_open(cagg_view_oid, AccessShareLock);
+	cagg_view_rules = cagg_view_rel->rd_rules;
+	Assert(cagg_view_rules && cagg_view_rules->numLocks == 1);
+
+	rule = cagg_view_rules->rules[0];
+	Ensure(rule->event == CMD_SELECT, "unexpected rule event %d for view", rule->event);
```
Comment (svenklemm) [inline src/ts_catalog/continuous_agg.c:1930]: this should not be ensure but proper error check

**Comment by maxhertrampf** [comment]
Thanks for adding nested caggs!

I have a quick question about this sentence:

> bucket width of the new continuous aggregate should be multiple of
the source continuous aggregate.

What exactly does "multiple" mean. Can I only built 24h on top of 1h (i.e. always same multiple)? Or can I also built 1day on top of 1 hour or 1 month on top of 1 day (variable multiple - of course respecting boundaries in the specified time zone)

**Review by konskov — APPROVED** [review]

File: src/ts_catalog/continuous_agg.c (hunk 1 above)
Comment (fabriziomello) [reply src/ts_catalog/continuous_agg.c]: Ooops... fixed!!!

File: tsl/test/sql/include/cagg_on_cagg_common.sql [hunk 2]
```diff
@@ -2,6 +2,8 @@
 -- Please see the included NOTICE for copyright information and
 -- LICENSE-TIMESCALE for a copy of the license.
 
+\set VERBOSITY default
```
Comment (svenklemm) [inline tsl/test/sql/include/cagg_on_cagg_common.sql:5]: why is this needed?

Comment (fabriziomello) [reply tsl/test/sql/include/cagg_on_cagg_common.sql:5]: want to see some `errdetail` in test output, specially in validations

File: tsl/test/expected/cagg_on_cagg_integer.out [hunk 3]
```diff
@@ -11,11 +11,14 @@
...
 \set BUCKET_WIDTH_3TH 'INTEGER \'10\''
--- Run tests
+--
+-- Run common tests
+--
 \ir include/cagg_on_cagg_common.sql
 -- This file and its contents are licensed under the Timescale License.
 -- Please see the included NOTICE for copyright information and
 -- LICENSE-TIMESCALE for a copy of the license.
+\set VERBOSITY default
```
Comment (mkindahl) [inline tsl/test/expected/cagg_on_cagg_integer.out:21]: This seems excessive. It will print the chunk names as well as other stuff and it is not particularly important for the tests below. If you want to print error details for a few particular cases, I suggest that you enable it for them alone and make sure that the chunk names are stable by, for example, creating a new hypertable.

File: tsl/test/expected/cagg_on_cagg_integer_dist_ht.out [hunk 4]
```diff
@@ -45,11 +45,14 @@ GRANT CREATE ON SCHEMA public TO :ROLE_DEFAULT_PERM_USER;
...
 \set BUCKET_WIDTH_3TH 'INTEGER \'10\''
--- Run tests
+--
+-- Run common tests
+--
 \ir include/cagg_on_cagg_common.sql
 -- This file and its contents are licensed under the Timescale License.
 -- Please see the included NOTICE for copyright information and
 -- LICENSE-TIMESCALE for a copy of the license.
+\set VERBOSITY default
```
Comment (mkindahl) [inline tsl/test/expected/cagg_on_cagg_integer_dist_ht.out:55]: Seems excessive. The error message might change as we add other tests and the detailed error messages are not important for the test.

File: tsl/test/expected/cagg_on_cagg_integer_dist_ht.out [hunk 5]
```diff
@@ -376,11 +387,13 @@ SELECT * FROM :CAGG_NAME_2TH_LEVEL ORDER BY bucket;
 
 -- DROP the 2TH level CAGG don't affect others
 DROP MATERIALIZED VIEW :CAGG_NAME_2TH_LEVEL;
-psql:include/cagg_on_cagg_common.sql:199: NOTICE:  drop cascades to table _timescaledb_internal._hyper_3_3_chunk
+psql:include/cagg_on_cagg_common.sql:201: NOTICE:  drop cascades to table _timescaledb_internal._hyper_3_3_chunk
```
Comment (mkindahl) [inline tsl/test/expected/cagg_on_cagg_integer_dist_ht.out:390]: I would actually disable notices here since they are not important for this test and will just generate false mismatches in the future.

File: tsl/test/sql/include/cagg_on_cagg_validations.sql [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+  SUM(temperature) AS temperature
+FROM conditions
+GROUP BY 1
+WITH NO DATA;
+
+--
+-- CAGG on CAGG (2th level)
+--
+\set ON_ERROR_STOP 0
+\echo :WARNING_MESSAGE
```
Comment (mkindahl) [inline tsl/test/sql/include/cagg_on_cagg_validations.sql:21]: Not sure what you need this for. It will be clear when you run the test and the error message will be in the output file, hence the expected error message after you have committed the PR. I think you can remove this line and the variable settings.

File: tsl/test/sql/include/cagg_on_cagg_common.sql (hunk 2 above)
Comment (mkindahl) [reply tsl/test/sql/include/cagg_on_cagg_common.sql:5]: Second that, it seems excessive. See previous comments.

**Review by mkindahl — COMMENTED** [review]
In general, this looks quite good, but I would consider the following:

- Using errors for cases that we think are not sensible might block users from interesting use-cases. I think a warning would be sufficient for these cases. Errors should be reserved for situations that will just not work.
- I think you should avoid enabling verbose message for entire tests and just increase and decrease the message level based on what is necessary. Maintaining tests with a lot of extraneous false mismatches takes extra time and risk missing actual failures.
- I think you're missing one test case.

File: tsl/src/continuous_aggs/create.c [hunk 7]
```diff
@@ -2507,7 +2590,11 @@ cagg_rebuild_view_definition(ContinuousAgg *agg, Hypertable *mat_ht)
 	Relation direct_view_rel = relation_open(direct_view_oid, AccessShareLock);
 	Query *direct_query = copyObject(get_view_query(direct_view_rel));
 	remove_old_and_new_rte_from_query(direct_query);
-	CAggTimebucketInfo timebucket_exprinfo = cagg_validate_query(direct_query, finalized);
+	CAggTimebucketInfo timebucket_exprinfo =
+		cagg_validate_query(direct_query,
+							finalized,
+							NameStr(agg->data.user_view_schema),
+							NameStr(agg->data.user_view_name));
```
[the remark this answers is not in the document]
Comment (fabriziomello) [reply tsl/src/continuous_aggs/create.c:2597]: To me seems a codecov mistake because I just change the `cagg_validate_query`  signature and in any case this code run just for old cagg version that we'll remove in a near future.

**Review by svenklemm — APPROVED** [review]

File: tsl/test/expected/cagg_on_cagg_integer.out (hunk 3 above)
Comment (fabriziomello) [reply tsl/test/expected/cagg_on_cagg_integer.out]: Fixed!

File: tsl/test/expected/cagg_on_cagg_integer_dist_ht.out (hunk 4 above)
Comment (fabriziomello) [reply tsl/test/expected/cagg_on_cagg_integer_dist_ht.out]: Fixed!

File: tsl/test/expected/cagg_on_cagg_integer_dist_ht.out (hunk 5 above)
Comment (fabriziomello) [reply tsl/test/expected/cagg_on_cagg_integer_dist_ht.out]: Fixed!

File: tsl/test/sql/include/cagg_on_cagg_common.sql (hunk 2 above)
Comment (fabriziomello) [reply tsl/test/sql/include/cagg_on_cagg_common.sql]: Fixed!

File: tsl/src/continuous_aggs/create.c [hunk 8]
```diff
@@ -1226,30 +1232,104 @@ cagg_validate_query(const Query *query, const bool finalized)
...
+		{
+			ereport(ERROR,
+					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+					 errmsg("cannot create continuous aggregate with fixed-width bucket on top of "
+							"one using variable-width bucket"),
+					 errdetail("Continuous aggregate with a fixed time bucket width (e.g. 61 days) "
+							   "cannot be created on top of one using variable time bucket width "
+							   "(e.g. 1 month).\n"
+							   "The variance can lead to the fixed width one not being a multiple "
+							   "of the variable width one.")));
```
[the remark this answers is not in the document]
Comment (fabriziomello) [reply tsl/src/continuous_aggs/create.c:1292]: We discussed a lot already about it (see the issue https://github.com/timescale/product/issues/337) and we decided to be more restrictive in this first version and perhaps relax it (or make it configurable) in the future.

File: tsl/src/continuous_aggs/create.c [hunk 9]
```diff
@@ -1226,30 +1232,104 @@ cagg_validate_query(const Query *query, const bool finalized)
...
+		{
+			ereport(ERROR,
+					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+					 errmsg("cannot create continuous aggregate with incompatible bucket width"),
+					 errdetail("Time bucket width of \"%s.%s\" should be greater than the time "
+							   "bucket width of \"%s.%s\".",
+							   cagg_schema,
+							   cagg_name,
+							   NameStr(cagg_source->data.user_view_schema),
+							   NameStr(cagg_source->data.user_view_name))));
```
[the remark this answers is not in the document]
Comment (fabriziomello) [reply tsl/src/continuous_aggs/create.c:1314]: Make sense add the bucket sizes. Will add!!!

File: tsl/src/continuous_aggs/create.c [hunk 10]
```diff
@@ -1226,30 +1232,104 @@ cagg_validate_query(const Query *query, const bool finalized)
...
+		{
+			ereport(ERROR,
+					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+					 errmsg("cannot create continuous aggregate with incompatible bucket width"),
+					 errdetail("Time bucket width of \"%s.%s\" should be multiple of the time "
+							   "bucket width of \"%s.%s\".",
+							   cagg_schema,
+							   cagg_name,
+							   NameStr(cagg_source->data.user_view_schema),
+							   NameStr(cagg_source->data.user_view_name))));
```
[the remark this answers is not in the document]
Comment (fabriziomello) [reply tsl/src/continuous_aggs/create.c:1328]: Same comment as before!!! It was already discussed and for this version we'll be more restrictive!

File: tsl/test/sql/include/cagg_on_cagg_validations.sql (hunk 6 above)
Comment (fabriziomello) [reply tsl/test/sql/include/cagg_on_cagg_validations.sql:22]: Added this message just for make our life easier when reading/underastangint the output test file. IMHO leaving it as is will not hurt us too much.

File: tsl/src/continuous_aggs/create.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (fabriziomello) [reply tsl/src/continuous_aggs/create.c:1317]: Added those informastion as `DEBUG1` level to don't generate too much noise!

**Comment by fabriziomello** [comment]
> 
> What exactly does "multiple" mean? Can I only build 24h on top of 1h (i.e. always the same multiple)? Or can I also build 1 day on top of 1 hour or 1 month on top of 1 day (i.e. variable multiple (most days have 24 hours, some 23/25; months can have 28-31 days) - of course respecting boundaries in the specified time zone)?
> 

All use cases you mentioned are supported, but the following:
* `90 days` (fixed-width bucket) on top of `1 month` (variable-width bucket) will not work
* `7 days` on top of `3 days` (are not multiple) will not work, but `9 days` on top of `3 days` will work

**Comment by maxhertrampf** [comment]
> All use cases you mentioned are supported

Thank you, @fabriziomello !

The not-working cases you listed make sense.
